Core/clim-basic/ports.lisp

```diff
@@ -275,21 +272,33 @@
             (new-pointer-cursor (sheet-pointer-cursor new-pointer-sheet)))
 	(unless (eql old-pointer-cursor new-pointer-cursor)
 	  (set-sheet-pointer-cursor port (event-sheet event) new-pointer-cursor))))
-    ;; Maybe-update the pressed sheet.
-    (cond
-      ((and (typep event 'pointer-button-press-event)
-            (null pressed-sheet))
-       (setf (port-pressed-sheet port) new-pointer-sheet))
-      ((and (typep event 'pointer-button-release-event)
-            (not (null pressed-sheet)))
-       (setf (port-pressed-sheet port) nil))
-      ((and (typep event 'pointer-motion-event)
-            (not (null pressed-sheet))
-            (not (eql pressed-sheet new-pointer-sheet)))
-       (dispatch-event-copy pressed-sheet event)))
+    ;; Pressing pointer button over a sheet makes a sheet pressed and
+    ;; focused. Releasing the button sets the presed sheet to NIL (but
+    ;; focus remains). Unless pressed sheet is already a target of the
+    ;; motion event, event is duplicated and dispatched to it.
+    (typecase event
+      (pointer-button-press-event
+       (when (null pressed-sheet)
+         (setf (port-pressed-sheet port) new-pointer-sheet))
+       (setf (port-focused-sheet port) new-pointer-sheet))
+      (pointer-button-release-event
+       (when pressed-sheet
+         (setf (port-pressed-sheet port) nil)))
+      (pointer-motion-event
+       (when (and pressed-sheet (not (eql pressed-sheet new-pointer-sheet)))
+         (dispatch-event-copy pressed-sheet event))))
     ;; Distribute event to the innermost child.
     (dispatch-event-copy new-pointer-sheet event)))
 
+(defmethod distribute-event ((port basic-port) (event keyboard-event))
+  (if-let ((focused (port-keyboard-input-focus port)))
+    (dispatch-event-copy focused event)
+    (let ((sheet (event-sheet event)))
+      ;; SETF the focus to avoid further changes when we hover the
+      ;; pointer over different sheet.
+      (setf (port-keyboard-input-focus port) sheet)
```

Comment:
Maybe we should have a protocol to compute the initial focus and focus changes as a result of events. I'm suggesting this because I see two potential problems:
* It is not clear which sheet will get the keyboard focus before the user has clicked one and there is no way customize that
* There is no way to implement other focus policies ("focus follows mouse" would be an alternative model)

Reply:
Customization should be made in a distribute-event method specialization (basic-port represents "default" behavior). I.e if we have wanted to implement "focus follows mouse" for CLX port (putting aside for now the fac that we'd also have clx's default behavior for pressed sheets and grabbed sheets), it would be enough to put in port.lisp

```
(defmethod distribute-event ((port clx-port) event)
  (dispatch-event (event-sheet event) event))
```

of course this method can accomodate any number of customizations with regard to changing what is focused or pressed. I don't think we want to make it configurable at runtime (yet?), so I'd rather address this problem when there is a need for it.

That's regarding the second point.

Regarding the first point I do agree, this PR does not maintain frame focus (so I'll need to fix it before we merge the branch). If you click on the window "decoration bar" focus will remain in the old sheet even if it is in a different frame.

As of defaults they are not much of a problem in case of frames, because their loop calls with-input-focus which sets the focused sheet to the frame's query-io (overwriting whatever we would set as a "default" strategy. When I ponder a little with frame "saved" focus I'll leave another comment here with a proposed solution.

Reply:
This is how it currently looks like:
```
(defmethod (setf port-keyboard-input-focus) :around (sheet port)
  (let ((old-sheet (port-keyboard-input-focus port))
        ;; Result may not reflect the actual focus.
        (result (call-next-method))
        (new-sheet (port-keyboard-input-focus port)))
    (prog1 result
      (unless (eq old-sheet new-sheet)
        (note-input-focus-changed old-sheet nil)
        (note-input-focus-changed new-sheet t)))))
```